### PR TITLE
Feature/some kind of card

### PR DESCRIPTION
**Pull request by henrikhermansen**
**This PR should not be merged**
For now, it is here for discussing styling and component API.

This relates to #260

### CODE REVIEW

File: packages/ffe-cards-react/src/NewCard.js
Comment (henrikhermansen) [inline packages/ffe-cards-react/src/NewCard.js:36]: Should the card heading be some kind of <h*> semantically?
If so, this needs to be decided by the consumer to make sure it's correct in their hierarchy of headings.

File: packages/ffe-cards-react/src/NewCard.js
Comment (henrikhermansen) [inline packages/ffe-cards-react/src/NewCard.js:37]: As it is now we have no children prop, so I don't feel like this is really react-ish.

A possible solution could be to replace `heading` and `details` and put `children` inside the `ffe-new-card__body`-div. This would make it easier for consumers to mess up the intended design, even if we say "we expect children to be a heading and a paragraph" and then we style these as they should be.

Thoughts on this predicament?

**Comment by henrikhermansen** [comment]
**Some screenshots of the current design**

image
_Cards in a grid_

image
_Condensed cards in a grid_

image
_A condensed card with greyCharcoal_

image
_This grid has 3 columns. It gets a little weird, but not too bad, right before `@breakpoint-md`_

image
_With 3 columns, it's time to break it down to just 1 when it gets narrow. This is just below `@breakpoint-md`_

File: packages/ffe-cards/less/ffe-new-card.less
Comment (henrikhermansen) [inline packages/ffe-cards/less/ffe-new-card.less]: Should all cards be clickable?
If not, I suppose we only want the cursor on clickable cards.
This is potentially something the consumer could be responsible for.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (henrikhermansen) [inline packages/ffe-cards/less/ffe-new-card.less]: Should all cards have a hover effect?
Perhaps not the ones that aren't clickable?
I assume the answer to this will be the same for focus.

File: packages/ffe-cards-react/src/NewCard.js
Comment (selbekk) [reply packages/ffe-cards-react/src/NewCard.js:36]: i don't think so - it's not a heading as much as it is the text content of most cards.

File: packages/ffe-cards-react/src/NewCard.js
Comment (selbekk) [reply packages/ffe-cards-react/src/NewCard.js:37]: A component doesn't need to have children - in cases like these, they typically don't (since things need to be in a certain order.

An alternative could be to use render props with what's known as compound components:

```js
<NewCard>
  {({ Heading, Details }) => (
    <Fragment>
      <Heading element="h2">My heading</Heading>
      <Details>This is some details</Details>
    </Fragment>
  )}
</NewCard>
```

, but I think that's major overkill.

Strict APIs that makes sure the consumer doesn't abuse a component's intended usecase might even lead to better DX (developer experience) in the long term, because we know what to use what component for.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (selbekk) [reply packages/ffe-cards/less/ffe-new-card.less]: Most cards should probably be clickable. Therefore I think using a `<button />` as the default Element in the React implementation might be a good idea. This way, we get a lot for free! Remember to reset `appearance` if you do though.

File: packages/ffe-cards-react/src/NewCard.js
Comment (ivarni) [reply packages/ffe-cards-react/src/NewCard.js:36]: You can always accept either a `string` or a `node`

File: packages/ffe-cards-react/src/NewCard.js
Comment (ivarni) [reply packages/ffe-cards-react/src/NewCard.js:37]: Hard to answer that without knowing how this component will be used in 1 year's time ;)

**Review by selbekk — DISMISSED** [review]
I think this looks good tbh - but why do we need a new component? Isn't this just a revamp of the original card components?

Also, if we're adding this, I would like a comprehensive list of usecases (ref your questions about clickability) and when NOT to use it.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (ivarni) [reply packages/ffe-cards/less/ffe-new-card.less]: You already accept `element` as a prop, if it's supposed to be a link, perhaps the consumer can provide `"a"` as the value and the cursor should act accordingly (I'd assume).

File: packages/ffe-cards-react/src/NewCard.js
Comment (henrikhermansen) [reply packages/ffe-cards-react/src/NewCard.js:37]: That's a cool technique @selbekk ! I never considered you could return components for render props.
That's true @ivarni . I just hope we can agree on some way to do it now. Nothing stands in the way of breaking that in a year's time and doing these cards in a different way.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (henrikhermansen) [reply packages/ffe-cards/less/ffe-new-card.less]: I'll look a bit into this, and see how `button` and `a` will affect the box.

Comment (adidick-zz) [reply packages/ffe-cards/less/ffe-new-card.less]: As per design definition, ALL cards should be clickable. Anything else is just a box grouping a set of information.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (adidick-zz) [reply packages/ffe-cards/less/ffe-new-card.less]: Same again, all cards are clickable so should have a hover effect.

**Comment by henrikhermansen** [comment]
@selbekk I agree this component is extremely similar to `ProductCard`. The difference is the class and a couple of modifiers not available to product card.
Is there a good way to melt these components together? I don't think I could just overwrite the old one, as people might be using it the way it is now.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (henrikhermansen) [inline packages/ffe-cards/less/ffe-new-card.less]: One thought on the styling that makes out the basics of this box. Perhaps it should be extracted as a totally separate class? I know people are using custom boxes with the same basics, so it might be good to help facilitate those boxes.
The new React component here could simply apply both classes `.ffe-card-basics` and `.ffe-new-card` (or whatever  we call them).

File: packages/ffe-cards-react/src/NewCard.js
Comment (henrikhermansen) [reply packages/ffe-cards-react/src/NewCard.js:36]: If it shouldn't be a heading tag anyways, I think accepting only string is fine here.

File: packages/ffe-cards/less/ffe-new-card.less
Comment (henrikhermansen) [reply packages/ffe-cards/less/ffe-new-card.less]: Thanks for clearing that up, Adi.

I see I need to add `text-align: left` if button is used as element.

Also, if element is `"a"` I get the pointer on the entire box by default, but as they should all be clickable, I think we should keep the cursor styling.

**Comment by adidick-zz** [comment]
Question: 
Does the naming of a card have to reflect it's use, eg. "ProductCard", or could we simplify it to just having one card that has multiple variations, depending on what it's going to be used for and where?

For example, product card, would just be "card with icon, title and subtext" from these examples. 
image

Or its it that in the implementation the component itself "product card" needs a name, and then could potentially take in use "card with icon, title and subtext" as it's master template.

I'm working with cards now in the sketch design file so wonder whats the best way to document them here as well.

**Comment by henrikhermansen** [comment]
@adidick I hope I fully understood your question, and I believe the answer is "a bit of both".

The way I see it now we could perhaps approach this in a similar way we've done with buttons.

We could have a base card class (which simply is the frame with hover and focus) and a component (which simply takes children). This will be used as a base for the other cards, and could be used when something completely custom is needed.

With this we could add specific styling and components for each type of card. Here, we should try to do some grouping. For example your two last cards are extremely similar, but they are clearly different from the cards above them. The way I see it the two last cards are the same, they are just variations on which types of they text they contain. The same goes for the three first cards. They are they same, give or take an image and a details text.
This will come down to balancing having too complex classes and components, and having too many classes and components.
The complexity of all this depends on how many variations of cards designers can come up with.. :smile: 

To better answer your questions: the naming of a card doesn't _have_ to reflect its use, but it might help to understand where, how and why it should be used. At the same time, individual cards needs a name. Like I said, we should balance complexity and diversity, so we'll have a certain number of cards, each with a certain set of variations.

Any thoughts on this @selbekk @ivarni ?

**Comment by antidecaf** [comment]
I agree with this approach from a styling perspective. All cards share quite a few base properties, and the variations of internal layouts in each card probably shouldn't have to be _that_ complex (?).

```
.ffe-base-card {
    // common styles (including a default layout)
    
    &--modifier {
        // text-align right or whatever other necessary modification
    }
}
```

**Comment by adidick-zz** [comment]
@henrikhermansen 
Nicely clarified. 
You've pretty much summed up what I was alluding to. Simple, structured, common sense structure that is both flexible but, as you said, helps guide both designers and developers to use the right card in the right situation. 
With that in mind it's maybe worth taking a bit of an audit of all of the cards and their uses (currently in prod and in working design sketches) and trying to find some natural groupings and uses so that we can start looking at how to best categorise the different types.
There will certainly be some cards that are used more often than others, eg. "navigation card", "product card", so once we start gathering the variations and the context in which they're used it should be relatively easy to start categorising creating the different types in a family.

**Comment by henrikhermansen** [comment]
@adidick thanks :blush:
We should definately go through all the cards to get a good overview before we implement them all.

I suggest I proceed with this PR by creating a basic class and component, as well as an initial implementation of the type of card I have created. If we do this implementation well it can serve as a guideline for implementing the rest of the cards later.

When we do the broader implementation of cards, we might want to consider removing the product card and link card which exists today. Perhaps these should be marked as deprecated in this PR as well?

**Comment by ivarni** [comment]
I thought the discussion in the HipChat room concluded that not every app will nessescarily get the grey background?

**Comment by ivarni** [comment]
> [11:32 AM] Adrian William Dick: Felles strategi? Hva er dette?...
[11:33 AM] Adrian William Dick: Det er kanskje at flere tenker sånn men ift en ordentlig strategi er jeg usikker på.
[11:35 AM] Adrian William Dick: Det er noen designere på bank som har sett på variasjoner av sand og lys grå som er litt varmere og at vi tenker kommer til å erstatte noen av disse eksisterende farger som finnes. Men jeg er litt usikker hva status er på denne saken.

Seems premature to deprecate the old ones unless we plan to use the new ones on white backgrounds as well

**Comment by henrikhermansen** [comment]
@ivarni is your question about grey background directed to my suggestion of deprecation? Or was it something else?

**Comment by ivarni** [comment]
I was under the impression that the new cards won't play nice with a white background, but if that assumption is wrong and we can move to the new ones, then sure.

But then we should look at if the new ones also fit the usecases we already have for the old.

Or we could just hold off deprecating anything :)

**Comment by henrikhermansen** [comment]
Hmm I'm not sure, and I haven't tried. I can have a look at how it looks tomorrow.
Are the new cards supposed to be fine on white background @adidick ?

I'm in no rush to deprecate anything, I was just asking :)

**Comment by adidick-zz** [comment]
Sorry, to clarify, the new cards MUST work on white backgrounds as well, otherwise they're of no use and the last thing we want to do is have to create a custom mongrel card that works only on white. The new card design with shadow works on white but is maybe not the best it could be in terms of accessibility, so that's something we'll just have to tweak as we go along. 

What I was thinking is that we edit the existing card style (white box with grey border) to the new style (white box, no border, drop shadow) and then tweak the shadow accordingly once we've seen how it looks/works in different situations. A white box will always look/work better on a coloured background because it'll have a better contrast than if it was on a white background, so the main thing to do here is make sure that the new card with shadow is good enough on a white background.

**Comment by adidick-zz** [comment]
@ivarni In regards to use cases, I think we can make a hypothesis that the new card will work on white (it should, it's really just a question of tweaking the contrast so it still looks good visually but is also accessible), and then hopefully use-cases will show up along the way to show us whether that is true or not. If not, we adjust the new design.

**Comment by ivarni** [comment]
OK, fair enough. Here's a few usecases I found in one of our apps now

image
This does not seem to be supported by the new implementation, so if we're going to keep having that status banner we'll need a new design for these or the new ones need to support it.

image
These seem like we can just use the new design directly as long as it looks nice with just the heading.

image
These aren't really cards per se, but they follow a similar pattern with the bordered box approach.

image
Here's a wider variant we're also using.

These are the ones I could find that could be (or already are) classified as cards. There might be others as well.

**Comment by henrikhermansen** [comment]
I've split the styling into a base class and a class for the card I've made.
Because of the size of the styling of our existing cards, I suggest we re-write them using the new base class. This should be a separate task imo.

@ivarni I see you have a nice selection of cards in your apps. I suppose it's a job for the designers to see if all of these should be commonly available cards, or if some are too specific, and should be local in your apps. For this you should be able to use the base class and component I've split out.

New commits are coming later today or tomorrow :)

**Comment by ivarni** [comment]
Having access to some kind of baseclass with the border would be very nice :)

**Comment by henrikhermansen** [comment]
New commits, ready for some feedback!

No need to mention I need to write tests. I know..
Edit: Coveralls is giving me a hard time over this already :roll_eyes:

**Comment by henrikhermansen** [comment]
This is what the cards currently look like

image
_GenericCard doesn't have a max-width or min-height_

image
image
_NewCard does not support the field "Kortnavn" from @adidick's sketches
It does look like his and these are the same type of card,
so should I perhaps go ahead and add that field?_

**Review by ivarni — DISMISSED** [review]
This LGTM, though I'd love to see a better name for this. And some tests. :trollface:

**Comment by henrikhermansen** [comment]
Tests are added. Still need a proper name for `NewCard` before this can be merged.

**Comment by antidecaf** [comment]
@henrikhermansen Any news on this?

**Comment by henrikhermansen** [comment]
There is news! I have resumed work on this.

**Comment by henrikhermansen** [comment]
I'll close the PR for now so I can make commits while I work, without notifying you all every time.

**Pull request closed without merging**
